File: examples/python/portpulse.py
"""
Minimal sample client for PortPulse API (for demos/tests).
No external deps except 'requests'. MIT-like usage.
"""
from __future__ import annotations
import os, typing as t, requests

class PortPulseClient:
    def __init__(self, api_key: str, base_url: str="https://api.useportpulse.com", timeout: int=15):
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
        self.session.headers.update({"X-API-Key": api_key, "Accept": "application/json"})
        self.timeout = timeout

    def health(self) -> dict:
        r = self.session.get(f"{self.base_url}/v1/health", timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def trend(self, unlocode: str, days: int = 7, fields: t.Optional[str] = None) -> dict:
        params = {"days": days}
        if fields:
            params["fields"] = fields
        r = self.session.get(f"{self.base_url}/v1/ports/{unlocode}/trend", params=params, timeout=self.timeout)
        r.raise_for_status()
        return r.json()

    def trend_csv(self, unlocode: str, days: int = 7, etag: t.Optional[str] = None) -> tuple[int, t.Optional[str], t.Optional[str]]:
        """
        Returns (status, csv_or_none, etag_or_none). If 304, csv is None.
        """
        headers = {"Accept": "text/csv"}
        if etag:
            headers["If-None-Match"] = etag
        r = self.session.get(
            f"{self.base_url}/v1/ports/{unlocode}/trend",
            params={"days": days, "format": "csv"},
            headers=headers, timeout=self.timeout,
        )
        if r.status_code == 304:
            return 304, None, etag
        r.raise_for_status()
        return 200, r.text, r.headers.get("ETag")

File: examples/js/portpulse.mjs
// Minimal sample client for PortPulse API (ESM). No deps.
export class PortPulseClient {
  constructor({ apiKey, baseUrl = "https://api.useportpulse.com", timeoutMs = 15000 } = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.timeoutMs = timeoutMs;
  }
  async _fetch(path, { params = {}, headers = {}, accept = "application/json" } = {}) {
    const url = new URL(this.baseUrl + path);
    for (const [k, v] of Object.entries(params)) if (v != null) url.searchParams.set(k, v);
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
    const res = await fetch(url, { headers: { "X-API-Key": this.apiKey, "Accept": accept, ...headers }, signal: ctrl.signal });
    clearTimeout(timer);
    if (res.status >= 400 && res.status !== 304) throw new Error(`HTTP ${res.status}: ${await res.text().catch(()=> "")}`);
    return res;
  }
  async health() {
    const r = await this._fetch("/v1/health");
    return r.json();
  }
  async trend(unlocode, { days = 7, fields } = {}) {
    const r = await this._fetch(`/v1/ports/${unlocode}/trend`, { params: { days, ...(fields ? { fields } : {}) } });
    return r.json();
  }
  async trendCsv(unlocode, { days = 7, etag } = {}) {
    const r = await this._fetch(`/v1/ports/${unlocode}/trend`, {
      params: { days, format: "csv" },
      headers: etag ? { "If-None-Match": etag } : {},
      accept: "text/csv"
    });
    if (r.status === 304) return { status: 304, csv: null, etag };
    return { status: 200, csv: await r.text(), etag: r.headers.get("etag") };
  }
}
